Eval.hs

```diff
@@ -444,6 +452,22 @@ isIdPair :: Val -> Bool
 isIdPair VIdPair{} = True
 isIdPair _         = False
 
+idComp :: Val -> Val -> Val -> Val -> Val -> Val -> Val
+idComp a u v w pId qId = case (pId, qId) of
+  (VIdPair p ps, qId) | eps `member` ps -> qId
+  (pId, VIdPair q qs) | eps `member` qs -> pId
```

Comment:
Are these two cases really needed? Isn't it possible to get the next case to cover them?